core/src/main/java/io/opencensus/stats/Measure.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2016, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.stats;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.Function;
+import io.opencensus.internal.StringUtil;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * Measure.
+ */
+@Immutable
+public abstract class Measure {
+
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  /**
+   * Applies the given match function to the underlying data type.
+   */
+  public abstract <T> T match(Function<DoubleMeasure, T> p0, Function<LongMeasure, T> p1);
+
+  /**
+   * Name of measure, as a {@code String}.
+   */
+  public abstract String getName();
+
+  /**
+   * Detailed description of the measure, used in documentation.
+   */
+  public abstract String getDescription();
+
+  /**
+   * The units in which {@link Measure} values are measured.
+   *
+   * <p>The grammar for a unit is as follows:
+   *     Expression = Component { "." Component } { "/" Component } ;
+   *     Component = [ PREFIX ] UNIT [ Annotation ] | Annotation | "1" ;
+   *     Annotation = "{" NAME "}" ;
+   * For example, string “MBy{transmitted}/ms” stands for megabytes per milliseconds, and the
+   * annotation transmitted inside {} is just a comment of the unit.
+   */
+  public abstract String getUnit();
+
+  @Immutable
+  @AutoValue
```

Comment:
AFAIK classes with AutoValue needs to be abstract and cannot be final, and the best we can do is to make the constructor package-private to prevent it from being extended anywhere else. For examples [DistributionView](https://github.com/census-instrumentation/opencensus-java/blob/master/core/src/main/java/io/opencensus/stats/View.java#L52). @sebright do you know a better way to fix this?